dvc/stage/run.py

```diff
@@ -145,48 +170,109 @@ def run_stage(
-        stage,
-        proc.pid,
-    )
-    done = threading.Event()
-    monitor_thread = threading.Thread(
-        target=_checkpoint_run,
-        args=(stage, callback_func, done, proc, killed),
-    )
-
-    try:
-        monitor_thread.start()
-        yield monitor_thread
-    finally:
-        done.set()
-        monitor_thread.join()
-
-
-def _checkpoint_run(stage, callback_func, done, proc, killed):
-    """Run callback_func whenever checkpoint signal file is present."""
-    signal_path = os.path.join(stage.repo.tmp_dir, CHECKPOINT_SIGNAL_FILE)
+@dataclass
+class MonitorConfig:
+    name: str
+    stage: "Stage"  # type: ignore[name-defined] # noqa: F821
+    task: Callable
+    proc: subprocess.Popen
+    done: threading.Event
+    killed: threading.Event
+    signal_path: str
+    AWAIT: ClassVar[float] = 1.0
+
+
+class Monitor(ABC):
+    def __init__(self, config: MonitorConfig):
+        self.config = config
+        self.monitor_thread = threading.Thread(
+            target=_monitor_loop, args=(self.config,),
+        )
+
+    def __enter__(self):
+        logger.debug(
+            "Monitoring stage '%s' with cmd process '%d'",
+            self.config.stage,
+            self.config.proc.pid,
+        )
+        self.monitor_thread.start()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.config.done.set()
+        self.monitor_thread.join()
+
+    @property
+    @abstractmethod
+    def SIGNAL_FILE(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def error_cls(self):
+        raise NotImplementedError
+
+    def signal_path(self, stage):
+        return os.path.join(stage.repo.tmp_dir, self.SIGNAL_FILE)
+
+
+class CheckpointMonitor(Monitor):
+    SIGNAL_FILE = "DVC_CHECKPOINT"
+    error_cls = CheckpointKilledError
+
+    def __init__(self, stage, callback_func, proc):
+        super().__init__(
+            MonitorConfig(
+                "checkpoint",
+                stage,
+                functools.partial(_run_callback, stage, callback_func),
+                proc,
+                threading.Event(),
+                threading.Event(),
+                self.signal_path(stage),
+            )
+        )
+
+
+class LiveMonitor(Monitor):
+    SIGNAL_FILE = "DVC_LIVE"
+    error_cls = LiveKilledError
+
+    def __init__(self, out, proc):
+        super().__init__(
+            MonitorConfig(
+                "live",
+                out.stage,
+                functools.partial(create_summary, out),
+                proc,
+                threading.Event(),
+                threading.Event(),
+                self.signal_path(out.stage),
+            )
+        )
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # make sure all data is visualied after training
+        self.config.task()
+        super().__exit__(exc_type, exc_val, exc_tb)
+
+
+def _monitor_loop(config: MonitorConfig):
```

Comment:
One general design question, does it make sense for us to have separate threads watching for both the checkpoint signal files and the dvclive signal files? Or should we have one monitor thread that checks for multiple signals (and runs their callbacks) sequentially like

```
while True:
    for signal_path, task in signals_to_watch:
        if os.path.exists(signal_path):
            try:
                task()
            ...
```

I guess the real question is would running both callbacks at the same time in parallel threads potentially cause any issues?

Reply:
When the user creates the checkpoint signal, their code is supposed to block so that we know the user won't be touching the DVC workspace until we signal them to continue (by removing the signal file). This way we know that we can do whatever is needed in the workspace to create the checkpoint commits safely, and I'm assuming that dvclive works the same way?

In the event that a user creates both a checkpoint and dvclive signal file at the same time, I'm not sure that it is safe for dvclive to be doing anything while the workspace may be in the state where we are moving HEAD around and creating git commits during checkpoint creation

Reply:
@pmrowla Your proposition makes more sense, Ill introduce necessary changes

Reply:
Redesigned so that we have single monitor with multiple tasks provided as needed, depending on stage content.